### ISSUE TITLE
[cppcoreguidelines-special-member-functions] Spiritually false positive for `operator=(T)`

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#76064]
Consider the example (https://godbolt.org/z/fPP6cnbYv):
```cpp
struct MyClass {
    MyClass(const MyClass&);
    MyClass(MyClass&&);
    ~MyClass();
    MyClass& operator=(MyClass other) noexcept;
};
```
```none
<source>:1:8: warning: class 'MyClass' defines a destructor, a copy constructor, a copy assignment operator and a move constructor but does not define a move assignment operator [cppcoreguidelines-special-member-functions]
    1 | struct MyClass {
      |        ^
```
Technically, clang-tidy is right, and `MyClass` does not have a move assignment operator.

However, this is spiritually a false positive because `operator=(MyClass)` accepts both rvaues and lvalues and is acting as a "unified" assignment operator. clang-tidy should not emit a diagnostic in the event that the copy assignment operator has a by-value parameter, or this behavior should be configurable somehow.

See [Stack Overflow/clang-tidy does not recognize unifying assignment operator as move assigment](https://stackoverflow.com/q/77692186/5740428) for more discussion.

**Comment by overmighty** [comment]
[Rule C.21 of the C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all) is about defining "either all or none of the copy/move/destructor functions" specifically, not just functions that can be called with both rvalues and lvalues.

**Comment by Eisenwave** [comment]
@overmighty if you use the copy-and-swap idiom, it's not actually possible to follow this idiom.
- You cannot declare `operator=(T&&) = default` since you're doing manual resource management, and this would be an ambiguous overload even if it did work.
- You cannot declare `operator=(T&&) = delete` since once again, you get ambiguous overloads.

Since it's impossible to follow this rule in that specific case, there needs to be some extra lenience in clang-tidy.

**Comment by overmighty** [comment]
@Eisenwave Since this Clang-Tidy check was made to enforce C.21, I believe that it shouldn't make any exceptions, as C.21 doesn't state any. If there are only rare cases where you do not want this diagnostic to be emitted, then you can simply suppress it for each case like this:

```cpp
struct MyClass { // NOLINT(cppcoreguidelines-special-member-functions)
    MyClass(const MyClass&);
    MyClass(MyClass&&);
    ~MyClass();
    MyClass& operator=(MyClass other) noexcept;
};
```

or:

```cpp
// NOLINTNEXTLINE(cppcoreguidelines-special-member-functions)
struct MyClass {
    MyClass(const MyClass&);
    MyClass(MyClass&&);
    ~MyClass();
    MyClass& operator=(MyClass other) noexcept;
};
```

**Comment by Eisenwave** [comment]
@overmighty I think C.21 simply doesn't have an exception because no one thought of this case, or no one bothered to state it explicitly. I've opened an issue in CppCoreGuidelines; perhaps the exception will added and the decision on this issue will be obvious.
